docs/source/integrations/middleware.md

```diff
@@ -1,47 +1,158 @@
+Because it sets helpful defaults, this library is also less configurable than other Apollo Server libraries. Complex projects might eventually need to swap it out for `apollo-server-express` (this process is straightforward).
 
-This document lists the options supported by each package's `applyMiddleware` or `getMiddleware` or `createHandler` function. You can learn more about how these options work in [the `ApolloServer` reference](../api/apollo-server/#framework-specific-middleware-function).
+</td>
+</tr>
 
-## `apollo-server`
+<tr>
+<td>
 
-`apollo-server` is the "batteries-included" GraphQL server. When you use it, you don't have to think about web frameworks or URLs or middleware, and its one main entry point (`listen`) is a nice `async` function rather than an old-school callback-based function.
+`apollo-server-express`
 
-```
-$ npm install apollo-server graphql
+</td>
+<td>
+
+This library enables you to attach Apollo Server to an [Express](https://expressjs.com/) server.
+
+The `apollo-server` library uses this library under the hood. Projects that start by using `apollo-server` sometimes need to swap it out for `apollo-server-express` when additional configuration is required.
+
+</td>
+</tr>
+
+<tr>
+<td>
+
+Serverless-specific packages
+
+</td>
+<td>
+
+The following libraries are available for running in a particular serverless environment:
+
+* `apollo-server-lambda` (AWS Lambda)
+* `apollo-server-cloud-functions` (Google Cloud Functions)
+* `apollo-server-azure-functions` (Azure Functions)
+* `apollo-server-cloudflare` (Cloudflare)
+
+Use each of these libraries if and only if you're running your server in the corresponding service.
+
+</td>
+</tr>
+
+
+<tr>
+<td>
+
+Framework-specific packages
+
+</td>
+<td>
+
+The following libraries are available for integrating Apollo Server with a particular Node.js framework:
+
+
+* `apollo-server-fastify` ([Fastify](https://www.fastify.io/))
+* `apollo-server-hapi` ([hapi](https://hapi.dev/))
+* `apollo-server-koa` ([Koa](https://koajs.com/))
+* `apollo-server-micro` ([Micro](https://www.npmjs.com/package/micro))
+
+Use each of these libraries if and only if you're integrating with the corresponding framework.
+
+</td>
+</tr>
+</tbody>
+</table>
+
+## Swapping out `apollo-server`
+
+After you get up and running with the "batteries-included" `apollo-server` package, you might want to configure its HTTP behavior in ways that this package doesn't support. For example, you might want to run some middleware before processing GraphQL requests, or you might want to serve other endpoints from the same server.
+
+In this case, we recommend you swap out `apollo-server` for `apollo-server-express` (unless you are confident that you want to use a different Node.js framework). This change requires only a few lines and has a minimal effect on your server's existing behavior (`apollo-server` uses `apollo-server-express` under the hood).
```

Comment:
Do we want to have a section here that directly shows how to swap out (ie what lines of code to write)? @trevor-scheer was saying that it felt like that was lacking and this seems like the right spot. (Not a regression in this sub-PR, shouldn't block merge)

Reply:
Definitely! I'm going to leave out of this PR and add to next